src/gui/notificationwidget.cpp

```diff
@@ -0,0 +1,147 @@
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "notificationwidget.h"
+#include "QProgressIndicator.h"
+#include "utility.h"
+
+#include <QPushButton>
+
+#include "ocsjob.h"
+
+namespace OCC {
+
+NotificationWidget::NotificationWidget(QWidget *parent) : QWidget(parent)
+{
+    _ui.setupUi(this);
+    _progressIndi = new QProgressIndicator(this);
+    _ui.horizontalLayout->addWidget(_progressIndi);
+}
+
+void NotificationWidget::setActivity(const Activity& activity)
+{
+    _myActivity = activity;
+
+    Q_ASSERT( !activity._accName.isEmpty() );
+    _accountName = activity._accName;
+
+    // _ui._headerLabel->setText( );
+    _ui._subjectLabel->setText(activity._subject);
+    if( activity._message.isEmpty()) {
+        _ui._messageLabel->hide();
+    } else {
+        _ui._messageLabel->setText(activity._message);
+    }
+    _ui._notifIcon->setPixmap(QPixmap(":/client/resources/bell.png"));
+    _ui._notifIcon->setMinimumWidth(64);
+    _ui._notifIcon->setMinimumHeight(64);
+    _ui._notifIcon->show();
+
+    QString tText = tr("Created at %1").arg(Utility::timeAgoInWords(activity._dateTime));
+    _ui._timeLabel->setText(tText);
+
+    // always remove the buttons
+    foreach( auto button, _ui._buttonBox->buttons() ) {
+        _ui._buttonBox->removeButton(button);
+    }
+    _buttons.clear();
+
+    // display buttons for the links
+    if( activity._links.isEmpty() ) {
+        // in case there is no action defined, do a close button.
+        QPushButton *b = _ui._buttonBox->addButton( QDialogButtonBox::Close );
+        b->setDefault(true);
+        connect(b, SIGNAL(clicked()), this, SLOT(slotButtonClicked()));
+        _buttons.append(b);
+    } else {
+        foreach( auto link, activity._links ) {
+            QPushButton *b = _ui._buttonBox->addButton(link._label, QDialogButtonBox::AcceptRole);
+            b->setDefault(link._isPrimary);
+            connect(b, SIGNAL(clicked()), this, SLOT(slotButtonClicked()));
+            _buttons.append(b);
+        }
+    }
+}
+
+Activity NotificationWidget::activity() const
+{
+    return _myActivity;
+}
+
+void NotificationWidget::slotButtonClicked()
+{
+    QObject *buttonWidget = QObject::sender();
+    int index = -1;
+    if( buttonWidget ) {
+        // find the button that was clicked, it has to be in the list
+        // of buttons that were added to the button box before.
+        for( int i = 0; i < _buttons.count(); i++ ) {
+            if( _buttons.at(i) == buttonWidget ) {
+                index = i;
+            }
+            _buttons.at(i)->setEnabled(false);
+        }
+
+        // if the button was found, the link must be called
+        if( index > -1 && _myActivity._links.count() == 0 ) {
```

Comment:
It's not that easy because the available buttons per notification is not fixed, it is part of the notification description. It has to be dynamic.